[quant][pt2e] Add reference representation for quantized add - relu

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105784
* #105783
* #105709
* #105708
* __->__ #105707

Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_add_relu

Although right now it is not really testing things since there is some problem with dynamo export
Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[quant][pt2e] Add reference representation for quantized add - relu

Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_add_relu

Although right now it is not really testing things since there is some problem with dynamo export
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/pt2e/representation/rewrite.py:
```diff
@@ -22,6 +23,37 @@
+    """
+    x_i32 = x_i8.to(torch.int32)
+    y_i32 = y_i8.to(torch.int32)
+    # TODO: use out_dtype op
```
not using this yet, there is some problem during export that need to resolved

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantized add - relu"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_add_relu

Although right now it is not really testing things since there is some problem with dynamo export
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vadimkantorov commented on torch/ao/quantization/pt2e/representation/rewrite.py:
```diff
@@ -22,6 +23,35 @@
+    y_i32 = out_dtype(torch.ops.aten.mul.Tensor, torch.int32, (y_i32 - y_zero_point), (y_scale / out_scale))
+    out_i32 = x_i32 + y_i32 + out_zero_point
+    out_i32 = torch.ops.aten.clamp(out_i32, out_zero_point)
+    quant_min = -128
```
quite strange to accept the `quant_min` as a method argument and then reset it like this. can this hide some bugs (if the passed `quant_min` is different from `-128`)?

jerryzh168 replied:
sorry, will update qmin/qmax and make it dtype agnostic as well. for add representation, it is not updated to use out_dtype yet, so I'll probably have a separate PR for that

jerryzh168 replied:
for dtype, can I do it a bit later to reduce the scope of change? it should be relatively straightforward

kimishpatel replied:
> for dtype, can I do it a bit later to reduce the scope of change? it should be relatively straightforward

Is this related to my comment? If it is straightforward and easy to include in this diff, please do so. Otherwise go ahead.

jerryzh168 replied:
this is related to some helper functions we added later in the stack (the replace literal with placeholder function) so I plan to have a separate PR after all these are done to include this support

kimishpatel commented on torch/ao/quantization/pt2e/representation/rewrite.py:
```diff
@@ -22,6 +23,35 @@
+    out_i32 = torch.ops.aten.clamp(out_i32, out_zero_point)
+    quant_min = -128
+    quant_max = 127
+    out_i8 = torch.ops.aten.clamp(out_i32, quant_min, quant_max).to(torch.int8)
```
You çån just change this to 
`torch.ops.aten.clamp(out_i32, out_zero_point, quant_max)...` since out_zero_point >= quant_min

kimishpatel reviewed (changes requested):
Left some comments.

Also concerned that tests are not really doing anything.

At high level, we should also move to registration API where decompositions are registered.

kimishpatel commented:
Also, we should really consider refactoring tests. Specifically representation tests can be separate.

jerryzh168 commented:
> 

yeah I can do this after this week, there are a few things that I need to finish before the end of this week

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantized add - relu"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_add_relu

Although right now it is not really testing things since there is some problem with dynamo export
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Add reference representation for quantized add - relu"


Summary:
Implementing reference representation for quantized ops we decided in https://docs.google.com/document/d/17h-OEtD4o_hoVuPqUFsdm5uo7psiNMY8ThN03F9ZZwg/edit#heading=h.ov8z39149wy8

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_representation_add_relu

Although right now it is not really testing things since there is some problem with dynamo export
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

andrewor14 reviewed (approved):
Stamping for now to unblock. Please follow up with Kimish's comments afterwards.

jerryzh168 commented:
@pytorchbot merge
